File: content_type.c
#include "content_type.h"
#include "global.h"
#include "textlist.h"
#include "indep.h"
#include "alloc.h"
#include <strings.h>

#ifndef NULL
#define NULL 0
#endif

static TextList* mimetypes_list = 0;
static struct ext_map** UserMimeTypes = 0;

// struct table2 {
//     char* item1;
//     char* item2;
// };
struct ext_map {
    const char* ext;
    const char* content_type;
};

static struct ext_map DefaultGuess[] = {
    { "html", "text/html" },
    { "htm", "text/html" },
    { "shtml", "text/html" },
    { "xhtml", "application/xhtml+xml" },
    { "gif", "image/gif" },
    { "jpeg", "image/jpeg" },
    { "jpg", "image/jpeg" },
    { "png", "image/png" },
    { "xbm", "image/xbm" },
    { "au", "audio/basic" },
    { "gz", "application/x-gzip" },
    { "Z", "application/x-compress" },
    { "bz2", "application/x-bzip" },
    { "tar", "application/x-tar" },
    { "zip", "application/x-zip" },
    { "lha", "application/x-lha" },
    { "lzh", "application/x-lha" },
    { "ps", "application/postscript" },
    { "pdf", "application/pdf" },
    { 0, 0 }
};

static struct ext_map* loadMimeTypes(char* filename)
{
    FILE* f;
    char *d, *type;
    int i, n;
    Str tmp;

    f = fopen(expandPath(filename), "r");
    if (f == NULL)
        return NULL;
    n = 0;
    while (tmp = Strfgets(f), tmp->length > 0) {
        d = tmp->ptr;
        if (d[0] != '#') {
            d = strtok(d, " \t\n\r");
            if (d != NULL) {
                d = strtok(NULL, " \t\n\r");
                for (i = 0; d != NULL; i++)
                    d = strtok(NULL, " \t\n\r");
                n += i;
            }
        }
    }
    fseek(f, 0, 0);
    struct ext_map* mtypes = New_N(struct ext_map, n + 1);
    i = 0;
    while (tmp = Strfgets(f), tmp->length > 0) {
        d = tmp->ptr;
        if (d[0] == '#')
            continue;
        type = strtok(d, " \t\n\r");
        if (type == NULL)
            continue;
        while (1) {
            d = strtok(NULL, " \t\n\r");
            if (d == NULL)
                break;
            mtypes[i].ext = Strnew_charp(d)->ptr;
            mtypes[i].content_type = Strnew_charp(type)->ptr;
            i++;
        }
    }
    mtypes[i].ext = NULL;
    mtypes[i].content_type = NULL;
    fclose(f);
    return mtypes;
}

void initMimeTypes(void)
{
    if (non_null(mimetypes_files))
        mimetypes_list = make_domain_list(mimetypes_files);
    else
        mimetypes_list = NULL;
    if (mimetypes_list == NULL)
        return;
    UserMimeTypes = New_N(struct ext_map*, mimetypes_list->nitem);
    int i = 0;
    for (TextListItem* tl = mimetypes_list->first; tl; i++, tl = tl->next)
        UserMimeTypes[i] = loadMimeTypes(tl->ptr);
}

static const char* guessContentTypeFromTable(struct ext_map* table, const char* filename)
{
    if (table == NULL)
        return NULL;
    const char* p = &filename[strlen(filename) - 1];
    while (filename < p && *p != '.')
        p--;
    if (p == filename)
        return NULL;
    p++;
    struct ext_map* t;
    for (t = table; t->ext; t++) {
        if (0 == strcasecmp(p, t->ext))
            return t->content_type;
    }
    return NULL;
}

const char* guessContentType(const char* filename)
{
    if (filename == NULL)
        return NULL;
    if (mimetypes_list == NULL)
        goto no_user_mimetypes;

    for (int i = 0; i < mimetypes_list->nitem; i++) {
        const char* ret;
        if ((ret = guessContentTypeFromTable(UserMimeTypes[i], filename)) != NULL)
            return ret;
    }

no_user_mimetypes:
    return guessContentTypeFromTable(DefaultGuess, filename);
}

File: w3m/content_type.zig
const std = @import("std");
const g = @import("global.zig");
const runtime = @import("runtime.zig");

var mimetypes_list: std.ArrayList([]const u8) = .initBuffer(&.{});

var UserMimeTypes: std.ArrayList([]ExtContentType) = .initBuffer(&.{});

const ExtContentType = struct {
    ext: [:0]const u8,
    content_type: [:0]const u8,
};

const DefaultGuess = [_]ExtContentType{
    .{ .ext = "html", .content_type = "text/html" },
    .{ .ext = "htm", .content_type = "text/html" },
    .{ .ext = "shtml", .content_type = "text/html" },
    .{ .ext = "xhtml", .content_type = "application/xhtml+xml" },
    .{ .ext = "gif", .content_type = "image/gif" },
    .{ .ext = "jpeg", .content_type = "image/jpeg" },
    .{ .ext = "jpg", .content_type = "image/jpeg" },
    .{ .ext = "png", .content_type = "image/png" },
    .{ .ext = "xbm", .content_type = "image/xbm" },
    .{ .ext = "au", .content_type = "audio/basic" },
    .{ .ext = "gz", .content_type = "application/x-gzip" },
    .{ .ext = "Z", .content_type = "application/x-compress" },
    .{ .ext = "bz2", .content_type = "application/x-bzip" },
    .{ .ext = "tar", .content_type = "application/x-tar" },
    .{ .ext = "zip", .content_type = "application/x-zip" },
    .{ .ext = "lha", .content_type = "application/x-lha" },
    .{ .ext = "lzh", .content_type = "application/x-lha" },
    .{ .ext = "ps", .content_type = "application/postscript" },
    .{ .ext = "pdf", .content_type = "application/pdf" },
};

fn loadMimeTypes(allocator: std.mem.Allocator, filename: []const u8) !?[]ExtContentType {
    const f = std.Io.Dir.cwd().openFile(runtime.io, filename, .{}) catch {
        return null;
    };
    defer f.close(runtime.io);

    var read_buf: [128]u8 = undefined;
    var reader = f.reader(runtime.io, &read_buf);

    var list: std.ArrayList(ExtContentType) = .initBuffer(&.{});
    defer list.deinit(allocator);
    while (true) {
        const d = reader.interface.takeDelimiterInclusive('\n') catch {
            break;
        };
        if (d[0] == '#')
            continue;
        var it = std.mem.splitAny(u8, d, &std.ascii.whitespace);
        const ext = it.next() orelse {
            continue;
        };
        while (it.next()) |content_type| {
            if (content_type.len == 0)
                continue;

            try list.append(allocator, .{
                .ext = allocator.dupeZ(u8, ext) catch @panic("OOM"),
                .content_type = allocator.dupeZ(u8, content_type) catch @panic("OOM"),
            });
        }
    }
    if (list.items.len > 0) {
        return try list.toOwnedSlice(allocator);
    } else {
        return null;
    }
}

export fn initMimeTypes() void {
    // update mimetypes_list
    mimetypes_list.clearRetainingCapacity();
    if (g.mimetypes_files) |files| {
        var it = std.mem.splitScalar(u8, std.mem.span(files), ',');
        while (it.next()) |_item| {
            const item = std.mem.trim(u8, _item, &std.ascii.whitespace);
            if (item.len > 0) {
                mimetypes_list.append(runtime.allocator, item) catch @panic("OOM");
            }
        }
    }

    UserMimeTypes.clearRetainingCapacity();
    for (mimetypes_list.items) |item| {
        if (loadMimeTypes(runtime.allocator, item) catch @panic("loadMimeTypes")) |list| {
            UserMimeTypes.append(runtime.allocator, list) catch @panic("OOM");
        }
    }
}

fn guessContentTypeFromTable(table: []const ExtContentType, path: []const u8) ?[:0]const u8 {
    const ext = std.fs.path.extension(path);
    if (ext.len == 0) {
        return null;
    }

    for (table) |item| {
        if (std.ascii.eqlIgnoreCase(ext[1..], item.ext)) {
            return item.content_type;
        }
    }

    return null;
}

pub export fn guessContentType(_src: [*c]const u8) [*c]const u8 {
    const src = std.mem.span(_src);
    if (src.len == 0)
        return null;

    for (UserMimeTypes.items) |table| {
        if (guessContentTypeFromTable(table, src)) |content_type| {
            return content_type.ptr;
        }
    }

    if (guessContentTypeFromTable(&DefaultGuess, src)) |content_type| {
        return content_type.ptr;
    }

    return null;
}
